subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_uart.c

```diff
@@ -0,0 +1,293 @@
+};
+
+static int request_expected_size(const struct ec_host_cmd_request_header *r)
+{
+	/* Check host request version */
+	if (r->prtcl_ver != 3) {
+		return 0;
+	}
+
+	/* Reserved byte should be 0 */
+	if (r->reserved) {
+		return 0;
+	}
+
+	return sizeof(*r) + r->data_len;
+}
+
+#define EC_HOST_CMD_UART_DEFINE(_name)                                                             \
+	static struct ec_host_cmd_uart_ctx _name##_hc_uart;                                        \
+	static struct ec_host_cmd_backend _name = {                                                \
+		.api = &ec_host_cmd_api,                                                           \
+		.ctx = &_name##_hc_uart,                                                           \
+	}
+
+/* Waiting time in microseconds to detect overrun */
+#define UART_OVERRUN_TIMEOUT_US 300
+/* Timeout after receiving first byte */
+#define UART_REQ_RX_TIMEOUT	K_MSEC(150)
+
+static void rx_timeout(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct ec_host_cmd_uart_ctx *hc_uart =
+		CONTAINER_OF(dwork, struct ec_host_cmd_uart_ctx, timeout_work);
+	int res;
+
+	switch (hc_uart->state) {
+	case USART_HOST_CMD_RECEIVING:
+		/* If state is receiving then timeout was hit due to underrun */
+		LOG_ERR("Request underrun detected");
+		break;
+	case USART_HOST_CMD_RX_OVERRUN:
+		/* If state is rx_overrun then timeout was hit because
+		 * process request was cancelled and extra rx bytes were
+		 * dropped
+		 */
+		LOG_ERR("Request overrun detected");
+		break;
+	case USART_HOST_CMD_RX_BAD:
+		/* If state is rx_bad then packet header was bad and process
+		 * request was cancelled to drop all incoming bytes.
+		 */
+		LOG_ERR("Bad packet header detected");
+		break;
+	default:
+		LOG_ERR("Request timeout mishandled, state: %d", hc_uart->state);
+	}
+
+	/* TODO: Workaround - reprogram DMA not to lose data at the end of buffer.
+	 * Remove it once the circular buffer is supported.
+	 */
+	res = uart_rx_disable(hc_uart->uart_dev);
+	res = uart_rx_enable(hc_uart->uart_dev, hc_uart->rx_ctx->buf,
+			     CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER, UART_OVERRUN_TIMEOUT_US);
+
+	hc_uart->state = USART_HOST_CMD_READY_TO_RX;
+}
+
+static void uart_callback(const struct device *dev, struct uart_event *evt, void *user_data)
+{
+	struct ec_host_cmd_uart_ctx *hc_uart = user_data;
+
+	switch (evt->type) {
+	case UART_RX_RDY:
+		uint8_t *evt_rx_buff = evt->data.rx.buf + evt->data.rx.offset;
+		uint8_t *rx_ctx_buff = hc_uart->rx_ctx->buf + hc_uart->rx_ctx->len;
+		size_t new_len = hc_uart->rx_ctx->len + evt->data.rx.len;
+
+		if (hc_uart->state == USART_HOST_CMD_READY_TO_RX) {
+			hc_uart->rx_ctx->len = 0;
+			hc_uart->state = USART_HOST_CMD_RECEIVING;
+			k_work_reschedule(&hc_uart->timeout_work, UART_REQ_RX_TIMEOUT);
+		} else if (hc_uart->state == USART_HOST_CMD_PROCESSING) {
+			LOG_ERR("USART HOST CMD ERROR: Received data while processing a command");
+			return;
+		}
```

Comment:
We should also return earlier on other `hc_uart` states, e.g. `USART_HOST_CMD_SENDING`, `USART_HOST_CMD_RX_BAD`

Reply:
Done.